api v1 - render errors with rabl

pitr-ch opened the pull request:
better detection of permission failure in model

Commit:
api v1 - render errors with rabl

better detection of permission failure in model
fix ApiConstraints
catch bad routes in api and return json

pitr-ch commented on app/models/authorization.rb:
```diff
@@ -30,7 +30,8 @@ def enforce_permissions operation
 
-    errors.add :base, "You do not have permission to #{operation} this #{klass}"
+    errors.add :base, message = "You do not have permission to #{operation} this #{klass}"
+    errors.add :permission, message
```
what do you thing about this one @ohadlevy ?
I did not want to remove the base error, there can be some dependencies on it. But I wanted to make it clear what type of error it is. 
see https://github.com/theforeman/foreman/pull/95/files#L0R19

pitr-ch replied:
never-mind, I've rewritten this, see https://github.com/theforeman/foreman/pull/95/files#diff-1

ohadlevy commented on app/controllers/api/base_controller.rb:
```diff
@@ -6,37 +6,42 @@ class BaseController < ActionController::Base
-      end
+    rescue_from StandardError, :with => lambda { |error|
+      Rails.logger.error "#{error.message} (#{error.class})\n#{error.backtrace.join("\n")}"
+      render :template => '/api/v1/errors/standard_error', :status => 500, :locals => { :error => error }
```
looks like we need a helper method to return the api version etc

pitr-ch replied:
fixed in baf2a47d41ef62e475c1db96242e3b4881d62a8b

ohadlevy commented on app/controllers/api/base_controller.rb:
```diff
@@ -6,37 +6,42 @@ class BaseController < ActionController::Base
+    protected
+
+    def process_resource_error(options = { })
+      resource = options[:resource] ||= get_resource || raise("No error to process")
```
wouldnt get resource raise before the last raise in this line ?

pitr-ch replied:
there should be: 

``` ruby
resource = options[:resource] || get_resource || raise("No resource to process")
```

I'll fix it in next commit

ohadlevy commented:
one thing that bothers me, is that we skip the controller permissions checks, and directly go to the model permissions checks.

foreman currently have a way to map permissions into web actions, and this basically means, that those permissions are only valid for the UI part of the site, and maybe we should consider mapping the api actions as well.

pitr-ch commented:
I've fixed the permission skiping in https://github.com/theforeman/foreman/pull/92/files, see #authorize. I think the pull-92 is implementing what you are describing.

pitr-ch commented:
> I'm not 100% sure, but if get_resource raise by it self, I dont think you'll ever get to the raise here.

You are right, fixed.

> nice catch, did you try to make sure it works? we currently have no tests for this code :)

I added some test for manager rights, it seems to be working ok :)

ohadlevy commented:
i mean the version matcher, not the permissions

pitr-ch commented:
>  out of the routes?

``` ruby
Rails.application.routes.routes.select { |r| r.verb == 'GET' && r.path =~ /\/api\// && r.path !~ /:id/ }
```

but it is not perfect, i look into it next week 

> rebase

yeah, it's not i'll do the squash and rebase

ohadlevy commented:
ok, please rebase :)

ohadlevy commented:
Please dont forget about https://github.com/theforeman/foreman/pull/95#issuecomment-6727357
